core/src/boot.py

```diff
@@ -3,14 +3,43 @@
 import storage.sd_salt
 from trezor import config, io, log, loop, res, ui, utils
 from trezor.pin import pin_to_int, show_pin_timeout
+from trezor.ui.confirm import HoldToConfirm
+from trezor.ui.text import Text
 
 from apps.common.request_pin import PinCancelled, request_pin
 from apps.common.sd_salt import SdProtectCancelled, request_sd_salt
 
 
+async def assert_sdcard() -> None:
+    sd = io.SDCard()
+    if not sd.present():
+        return
+
+    sd.power(True)
+    fs = io.FatFS()
+    try:
+        fs.mount()
+    except OSError:
+        text = Text("SD card", ui.ICON_WRONG, ui.RED)
+        text.normal("Unrecognized filesystem")
+        text.normal("found. Unplug the device")
+        text.normal("and remove the card or")
+        text.normal("you can format the card")
+        text.normal("losing all the data on it.")
+        dialog = HoldToConfirm(text, confirm="Format SD card")
+        await dialog
+        fs.mkfs()
+        fs.setlabel("TREZOR")
+    else:
+        fs.unmount()
+    sd.power(False)
+
+
 async def bootscreen() -> None:
     ui.display.orientation(storage.device.get_rotation())
 
+    await assert_sdcard()
```

Comment:
This means that Trezor will offer to format the SD card only if it's inserted before powering up. When the SD card is hotswappable, the the user can insert the SD card at any time, so:
1. User powers-up Trezor and unlocks it.
2. User tries to enable SD protect.
3. Trezor shows "Please insert your SD card ."
4. User inserts card and presses "Retry".
5. Trezor shows "There was a problem accessing your SD card." instead of offering to format it.

So I think the place where we should be checking for `fs.mount()` failure is https://github.com/trezor/trezor-firmware/blob/1480f5a25c5c40297de98790e4d6f59836d3c832/core/src/storage/sd_salt.py#L68.

Reply:
Agreed. Will rework the PR.